Replace Mobile.OnRails hack with dedicated OnRails traits

reaperrr opened the pull request:
While the new `OnRails` and `AttackOnRailsTurreted` are essentially hacky as well,
- they don't meddle with Common code (literally)
- they work better, because on bleed the gunboat often stops for a few ticks when attacking, while with this PR it works just like in the original game
- they allow to remove waypoints and lua triggers from some TD missions

Additionally, used this opportunity to replace the hacky camera actors in those missions with adjusted `RevealsShroud.ValidStances`.

reaperrr committed:
Move AttackPopupTurreted to Attack subfolder

reaperrr committed:
Add dedicated TDGunboat traits

And get rid of Mobile.OnRails hack.

reaperrr committed:
Adapt TD missions to TDGunboat refactor

abcdefg30 commented on OpenRA.Mods.Cnc/Traits/OnRails.cs:
```diff
@@ -0,0 +1,219 @@
+			init.Add(new FacingInit(Facing));
+		}
+
+		#region Implement IPositionable
```
We don't (usually) have those elsewhere, do we? So I think they should be removed here as well.

reaperrr replied:
Copy-paste from `Aircraft`, but I don't mind removing them.

abcdefg30 replied:
`[15:46:16] <pchote> we implicitly prefer explicit interface implementations over regions`, which probably means you should remove the regions and implement those interfaces explicitly.

reaperrr replied:
iirc I had some trouble making the interfaces explicit in the past (in Aircraft, at least) which is why I didn't touch this at first, but I guess if I run into that I can use subclassing to fix it.

pchote replied:
I think this is fine as-is.  Just remove the #regions.  Or don't.  I really don't think this is a big deal either way.

abcdefg30 reviewed (approved):
Looks good to me otherwise. 👍

reaperrr committed:
Implement damage-based slowdown on TD Gunboats

In line with original.

reaperrr commented:
Updated. Fixed upgrade rule comment, removed regions, added DamageState-based slowdown to TD Gunboat (they're singleplayer only, so it shouldn't matter if it makes them a little easier to kill).

pchote commented on OpenRA.Mods.Cnc/Traits/OnRails.cs:
```diff
@@ -0,0 +1,211 @@
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	public class OnRailsInfo : ITraitInfo, IPositionableInfo, IFacingInfo, IMoveInfo,
```
How many other mods are likely to need a unit that has exactly two artwork facings and moves left-right bouncing off the edge of the map?

I think it would be better to call this `TDGunboat` or similar so that we keep `OnRails` available for a less specific use case.

reaperrr replied:
Ok. Should I rename the Attack trait, too?

GraionDilach replied:
I'd rename that as well, yes.

reaperrr commented:
Updated.

obrakmann commented:
👍

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/34949820728cf18d98be9e200cb86327dafba4f8)
